python/hail/dataset.py

```diff
@@ -1019,181 +886,162 @@ def annotate_variants_expr(self, expr):
-        ...       'va.foo = table.foo',
-        ...       ['va.gene', 'if (va.score > 10) "Type1" else "Type2"']))
-
+        Annotate variants with the target in a GATK interval list file:
+        
+        >>> intervals = KeyTable.import_interval_list('data/exons2.interval_list')
+        >>> vds_result = vds.annotate_variants_table(intervals, root='va.exon')
+        
+        Annotate variants with all targets from matching intervals in a GATK interval list file:
+        
+        >>> intervals = KeyTable.import_interval_list('data/exons2.interval_list')
+        >>> vds_result = vds.annotate_variants_table(intervals, root='va.exons', product=True)
+        
+        Annotate variants using a UCSC BED file, marking each variant true/false for an overlap with any interval:
+        
+        >>> intervals = KeyTable.import_bed('data/file2.bed')
+        >>> vds_result = vds.annotate_variants_table(intervals, root='va.bed')
+        
         **Notes**
+        
+        This method takes as an argument a :class:`.KeyTable` object. Hail has defined join strategies
+        for tables keyed by Variant, Locus, or Interval. If the table is keyed by something else (gene ID, etc)
+        then the ``vds_key`` argument must be passed to describe the keys in the dataset to use for the join.
+        
+        **Join strategies:**
+                  
+        If the key is a ``Variant``, then a variant in the dataset will be annotated by finding a
+        complete match in the table. Be careful, however: ``1:1:A:T`` does not match ``1:1:A:T,C``, 
+        and vice versa. 
+        
+        If the key is a ``Locus``, then a variant in the dataset will be annotated by matching 
+        chromosome and position.
+        
+        If the key is an ``Interval``, then a variant in the dataset will be annotated by finding
+        an interval that contains the variant's chromosome and position.
+        
+        If the key is not one of the above three types or if another join strategy should be used
+        for these types, then the ``vds_key`` argument should be passed. This argument expects
+        a list of Hail expressions whose types match, in order, the table's key types. Note that 
+        using ``vds_key`` is slower than annotation with a standard key type.
 
-        ``expr`` has the following symbols in scope:
-
-          - ``va``: variant annotations
-          - ``table``: :py:class:`.KeyTable` value
-
-        each expression in the list ``vds_key`` has the following symbols in
+        Each expression in the list ``vds_key`` has the following symbols in
         scope:
 
           - ``v`` (*Variant*): :ref:`variant`
           - ``va``: variant annotations
+        
+        **The ``root`` and ``expr`` arguments**
+        
+        .. note::
+        
+            One of ``root`` or ``expr`` is required, but not both. 
+            
+        The ``expr`` parameter expects an annotation expression involving ``va`` (the existing 
```

Comment:
Are we calling these "annotation expressions"?  Maybe "annotation assignment" is better.